File: go/timehash.go
// Timehash
// Copyright (c) 2026-present NAVER Corp.
// Apache-2.0

package timehash

import (
	"fmt"
	"strconv"
)

// Default measures: [400, 100, 15, 5, 1] representing:
// - 400 minutes (4 hours)
// - 100 minutes (1 hour, treated as 60 minutes)
// - 15 minutes
// - 5 minutes
// - 1 minute
var defaultMeasures = []int{400, 100, 15, 5, 1}

// Timehash generates hierarchical hash keys for time ranges.
type Timehash struct {
	measures []int
}

// New creates a new Timehash instance with default measures.
func New() *Timehash {
	measures := make([]int, len(defaultMeasures))
	copy(measures, defaultMeasures)
	return &Timehash{measures: measures}
}

// NewWithMeasures creates a new Timehash instance with custom measures.
func NewWithMeasures(measures []int) *Timehash {
	m := make([]int, len(measures))
	copy(m, measures)
	return &Timehash{measures: m}
}

// GetQueryTerms generates query terms for a given time.
// hhmm: Time in hhmm format (e.g., 1720 for 17:20)
// Returns: List of hierarchical hash keys for the query time
func (th *Timehash) GetQueryTerms(hhmm int) []string {
	var terms []string

	if hhmm > 10000 {
		return terms
	}

	for _, measure := range th.measures {
		target := hhmm
		var measureVal int

		if measure >= 100 {
			target /= 100
			measureVal = measure / 100
		} else {
			target %= 100
			measureVal = measure
		}

		key := (target / measureVal) * measureVal
		term := th.twoDigitTerm(key)

		if len(terms) == 0 {
			terms = append(terms, term)
		} else {
			terms = append(terms, terms[len(terms)-1]+term)
		}
	}

	return terms
}

// GetIndexTerms generates index terms for a time range.
// fromTime: Start time in hhmm format (e.g., "1140")
// toTime: End time in hhmm format (e.g., "2100")
// Returns: List of hierarchical hash keys covering the time range
func (th *Timehash) GetIndexTerms(fromTime, toTime string) ([]string, error) {
	fromInt, err := strconv.Atoi(fromTime)
	if err != nil {
		return nil, fmt.Errorf("invalid fromTime: %w", err)
	}

	toInt, err := strconv.Atoi(toTime)
	if err != nil {
		return nil, fmt.Errorf("invalid toTime: %w", err)
	}

	var terms []string
	th.cover(&terms, 0, "", fromInt, toInt)
	return terms, nil
}

// twoDigitTerm converts a key to two-digit string representation.
func (th *Timehash) twoDigitTerm(key int) string {
	if key >= 100 {
		key /= 100
	}
	return fmt.Sprintf("%d%d", key/10, key%10)
}

// cover recursively generates keys for a time range.
func (th *Timehash) cover(terms *[]string, measureIndex int, parentKey string,
	fromTime, toTime int) {
	if measureIndex >= len(th.measures) {
		return
	}

	measure := th.measures[measureIndex]
	fromVal := fromTime
	toVal := toTime

	if measure < 100 {
		fromVal %= 100
		toVal %= 100
		if toVal == 0 {
			toVal = 60
		}
	}

	for i := fromVal / measure; i <= (toVal / measure); i++ {
		key := measure * i

		if fromVal <= key && key+measure <= toVal {
			*terms = append(*terms, parentKey+th.twoDigitTerm(key))
		} else if key < fromVal && fromVal < key+measure {
			end := key + measure
			if toVal < end {
				end = toVal
			}
			th.cover(terms, measureIndex+1,
				parentKey+th.twoDigitTerm(key),
				fromVal, end)
		} else if key < toVal && toVal < key+measure {
			th.cover(terms, measureIndex+1,
				parentKey+th.twoDigitTerm(key),
				key, toVal)
		}
	}
}

File: scala/src/main/scala/com/naver/timehash/Timehash.scala
/*
 * Timehash
 * Copyright (c) 2026-present NAVER Corp.
 * Apache-2.0
 */

package com.naver.timehash

/**
 * Timehash: Hierarchical time indexing for efficient business hours search.
 *
 * Timehash generates hierarchical hash keys for time ranges, enabling efficient
 * temporal filtering in search systems with minimal index overhead.
 */
class Timehash(private val measures: List[Int] = Timehash.DefaultMeasures) {
  /**
   * Generate query terms for a given time.
   *
   * @param hhmm Time in hhmm format (e.g., 1720 for 17:20)
   * @return List of hierarchical hash keys for the query time
   */
  def getQueryTerms(hhmm: Int): List[String] = {
    if (hhmm > 10000) {
      return List.empty
    }

    measures.foldLeft(List.empty[String]) { (terms, measure) =>
      val (target, measureVal) = if (measure >= 100) {
        (hhmm / 100, measure / 100)
      } else {
        (hhmm % 100, measure)
      }

      val key = (target / measureVal) * measureVal
      val term = twoDigitTerm(key)

      if (terms.isEmpty) {
        List(term)
      } else {
        terms :+ (terms.last + term)
      }
    }
  }

  /**
   * Generate index terms for a time range.
   *
   * @param fromTime Start time in hhmm format (e.g., "1140")
   * @param toTime End time in hhmm format (e.g., "2100")
   * @return List of hierarchical hash keys covering the time range
   */
  def getIndexTerms(fromTime: String, toTime: String): List[String] = {
    val fromInt = fromTime.toIntOption.getOrElse(0)
    val toInt = toTime.toIntOption.getOrElse(0)
    val terms = scala.collection.mutable.ListBuffer.empty[String]
    cover(terms, 0, "", fromInt, toInt)
    terms.toList
  }

  private def twoDigitTerm(key: Int): String = {
    val k = if (key >= 100) key / 100 else key
    s"${k / 10}${k % 10}"
  }

  private def cover(
    terms: scala.collection.mutable.ListBuffer[String],
    measureIndex: Int,
    parentKey: String,
    from: Int,
    to: Int
  ): Unit = {
    if (measureIndex >= measures.length) {
      return
    }

    val measure = measures(measureIndex)
    var fromVal = from
    var toVal = to

    if (measure < 100) {
      fromVal %= 100
      toVal %= 100
      if (toVal == 0) {
        toVal = 60
      }
    }

    val start = fromVal / measure
    val end = toVal / measure

    for (i <- start to end) {
      val key = measure * i

      if (fromVal <= key && key + measure <= toVal) {
        val term = twoDigitTerm(key)
        terms += (parentKey + term)
      } else if (key < fromVal && fromVal < key + measure) {
        val term = twoDigitTerm(key)
        val end = math.min(key + measure, toVal)
        cover(terms, measureIndex + 1, parentKey + term, fromVal, end)
      } else if (key < toVal && toVal < key + measure) {
        val term = twoDigitTerm(key)
        cover(terms, measureIndex + 1, parentKey + term, key, toVal)
      }
    }
  }
}

object Timehash {
  /**
   * Default measures: [400, 100, 15, 5, 1] representing:
   * - 400 minutes (4 hours)
   * - 100 minutes (1 hour, treated as 60 minutes)
   * - 15 minutes
   * - 5 minutes
   * - 1 minute
   */
  val DefaultMeasures: List[Int] = List(400, 100, 15, 5, 1)

  /**
   * Create a new Timehash instance with default measures.
   */
  def apply(): Timehash = new Timehash()

  /**
   * Create a new Timehash instance with custom measures.
   */
  def apply(measures: List[Int]): Timehash = new Timehash(measures)
}
